source/common/common/logger.cc

```diff
@@ -157,20 +157,29 @@ void DelegatingLogSink::setTlsDelegate(SinkDelegate* sink) { *tlsSink() = sink;
 
 SinkDelegate* DelegatingLogSink::tlsDelegate() { return *tlsSink(); }
 
-static Context* current_context = nullptr;
+ABSL_CONST_INIT static absl::Mutex current_context_mutex(absl::kConstInit);
+static Context* current_context ABSL_GUARDED_BY(current_context_mutex) = nullptr;
 
 Context::Context(spdlog::level::level_enum log_level, const std::string& log_format,
                  Thread::BasicLockable& lock, bool should_escape, bool enable_fine_grain_logging)
     : log_level_(log_level), log_format_(log_format), lock_(lock), should_escape_(should_escape),
-      enable_fine_grain_logging_(enable_fine_grain_logging), save_context_(current_context) {
-  current_context = this;
+      enable_fine_grain_logging_(enable_fine_grain_logging) {
+  {
+    absl::MutexLock context_lock(&current_context_mutex);
+    save_context_ = current_context;
+    current_context = this;
+  } // context_lock
   activate();
 }
 
 Context::~Context() {
-  current_context = save_context_;
-  if (current_context != nullptr) {
-    current_context->activate();
+  {
+    absl::MutexLock context_lock(&current_context_mutex);
+    current_context = save_context_;
```

Comment:
Should we first `ASSERT(current_context == this);` ? This stacked context discipline might behave strangely if you don't have strict nesting of the construct/destruct sequence so it might be better to catch a bad usage with an assert.

Reply:
Great call! in fact when I started working on this I had exactly that issue... I'll add an ASSERT with a comment to explain it.